### PR TITLE
feat: handle GitLab push rule violation

### LINKED ISSUE #8414: Handle push failures to GitLab project when push rules prevent pushing
**What would you like Renovate to be able to do?**

GitLab allow setting push rules that can restrict:
- branch name
- commit message

to follow organisation and/or project rules/agreement/policy.

This can cause push failures when Renovate bot is trying to changes to GitLab. As information of this failure is only in logs bot user is not aware of configuration problem. I think this push failure should generate configuration issue in similar way like it does for invalid config or inaccessible extends.

Here is log snippet when Renovate is trying to push onboarding MR and branch name doesn't match rules set in GitLab:
```json
{
"name":"renovate",
"hostname":"runner-72989761-project-23909422-concurrent-0",
"pid":11,
"level":20,
"logContext":"MvcbpsEvy",
"repository":"wiget/renovate-push-rule-issue",
"err":
{"task":
{"commands":["push","origin","renovate/configure:renovate/configure","--force","-u","--no-verify","--verbose","--porcelain"],"format":"utf-8"},
"message":
"Pushing to https://gitlab.com/wiget/renovate-push-rule-issue.git\nPOST git-receive-pack (626 bytes)\nremote: GitLab: Branch name does not follow the pattern '(feat|fix|doc)\\/*'\nerror: failed to push some refs to 'https://gitlab.com/wiget/renovate-push-rule-issue.git'\n",
"stack":"Error: Pushing to https://gitlab.com/wiget/renovate-push-rule-issue.git\nPOST git-receive-pack (626 bytes)\nremote: GitLab: Branch name does not follow the pattern '(feat|fix|doc)\\/*'\nerror: failed to push some refs to 'https://gitlab.com/wiget/renovate-push-rule-issue.git'\n\n    at GitExecutorChain.onFatalException (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:67:77)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:59:21)\n    at Generator.throw (<anonymous>)\n    at rejected (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:6:65)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)"},
"msg":"Error committing files","time":"2021-01-25T14:19:35.742Z",
"v":0
}
```

```
Error: Pushing to https://gitlab.com/wiget/renovate-push-rule-issue.git
POST git-receive-pack (626 bytes)
remote: GitLab: Branch name does not follow the pattern '(feat|fix|doc)\\/*'
error: failed to push some refs to 'https://gitlab.com/wiget/renovate-push-rule-issue.git'

    at GitExecutorChain.onFatalException (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:67:77)
    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:59:21)
    at Generator.throw (<anonymous>)
    at rejected (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:6:65)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

And for commit message:
```json
{
"name":"renovate",
"hostname":"runner-72989761-project-23909422-concurrent-0",
"pid":12,
"level":20,
"logContext":"TusqOzdsb",
"repository":"wiget/renovate-push-rule-issue",
"branch":"renovate/pin-dependencies",
"err":{"task":
{"commands":["push","origin","renovate/pin-dependencies:renovate/pin-dependencies","--force","-u","--no-verify","--verbose","--porcelain"],"format":"utf-8"},
"message":"Pushing to https://gitlab.com/wiget/renovate-push-rule-issue.git\nPOST git-receive-pack (653 bytes)\nremote: GitLab: Commit message does not follow the pattern 'TICKET-.*'\nerror: failed to push some refs to 'https://gitlab.com/wiget/renovate-push-rule-issue.git'\n",
"stack":"Error: Pushing to https://gitlab.com/wiget/renovate-push-rule-issue.git\nPOST git-receive-pack (653 bytes)\nremote: GitLab: Commit message does not follow the pattern 'TICKET-.*'\nerror: failed to push some refs to 'https://gitlab.com/wiget/renovate-push-rule-issue.git'\n\n    at GitExecutorChain.onFatalException (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:67:77)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:59:21)\n    at Generator.throw (<anonymous>)\n    at rejected (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:6:65)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)"},
"msg":"Error committing files","time":"2021-01-25T14:44:21.527Z",
"v":0
}
```

```
Error: Pushing to https://gitlab.com/wiget/renovate-push-rule-issue.git
POST git-receive-pack (653 bytes)
remote: GitLab: Commit message does not follow the pattern 'TICKET-.*'
error: failed to push some refs to 'https://gitlab.com/wiget/renovate-push-rule-issue.git'

    at GitExecutorChain.onFatalException (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:67:77)
    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:59:21)
    at Generator.throw (<anonymous>)
    at rejected (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:6:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```
**Did you already have any implementation ideas?**

### PR DESCRIPTION
**Pull request by wiget**
## Changes:

Handle GitLab push rule violation error and convert it to config validation error to inform user that Renovate bot configuration does not match GitLab project configuration in terms of branch name and/or commit message.

## Context:

Closes #8414

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/index.ts [hunk 1]
```diff
@@ -70,12 +71,31 @@ function checkForPlatformFailure(err: Error): void {
...
   ];
-  for (const errorStr of platformFailureStrings) {
+  for (const errorStr of externalHostFailureStrings) {
     if (err.message.includes(errorStr)) {
       logger.debug({ err }, 'Converting git error to ExternalHostError');
       throw new ExternalHostError(err, 'git');
     }
   }
+
+  const gitLabPushRuleFailureStrings = [
```
Comment (viceice) [inline lib/util/git/index.ts:81]: I think this should be more generic, so we can reuse it for other platforms.

**Review by viceice — COMMENTED** [review]
We should also have some simple tests for this.

File: lib/util/git/index.ts (hunk 1 above)
Comment (wiget) [reply lib/util/git/index.ts:81]: Will `pushRuleFailureStrings` be ok?

**Comment by wiget** [comment]
@viceice To be honest I don't know how to write tests using Jest.

Something like this?
```typescript
describe('checkForPlatformFailure`, () => {
 it('should handle push rule violation', () => {
   cont err = new Error(...);
   except(checkForPlatformFailure(err)).toThrowErrorMatchingSnapshot();
  });
});
```

File: lib/util/git/index.ts (hunk 1 above)
Comment (viceice) [reply lib/util/git/index.ts:81]: I think `configErrorStrings` is a better option. We can move some of the `externalHostFailureStrings` here.

* `TF401027:` 
* `remote: Invalid username or password`

**Comment by wiget** [comment]
`checkForPlatformFailure` is not exported (which is good because this function is implementation detail) so it can't be tested in `lib/util/git/index.spec.ts`. Any suggestion?

**Review by viceice — DISMISSED** [review]
OK, we can do without tests. Need to think howto testing this in future

**Comment by rarkins** [comment]
Is it tested on a real repo to confirm that an issue is created as expected?

**Comment by wiget** [comment]
@rarkins not yet. I will try to do this.

**Comment by wiget** [comment]
I can't run renovate on linux/arm64 and no access to other devices. I will run tests tomorrow on linux/amd64.

**Comment by wiget** [comment]
Two more changes was needed to correctly create issue.

Here you can see examples for created issues:
- [commit message case](https://gitlab.com/wiget/renovate-push-rule-issue/-/issues/1)
- [branch name case](https://gitlab.com/wiget/renovate-push-rule-issue/-/issues/2)

**Review by rarkins — APPROVED** [review]

**Pull request merged**